onnxscript/ir/_core.py

```diff
@@ -1977,8 +1978,54 @@ def insert_before(self, node: Node, new_nodes: Iterable[Node] | Node, /) -> None
         self._nodes.insert_before(node, new_nodes)
 
     def sort(self) -> None:
-        """Topologically sort the nodes in the graph."""
-        raise NotImplementedError("Not implemented yet")
+        """Perform a topological sort of the graph.
+
+        This method sorts the nodes of a graph to ensure that each node appears before any nodes it
+        points to, while maintaining the original order as much as possible.
+
+        Raises:
+            ValueError: If the graph contains a cycle, making topological sorting impossible.
+        """
+        # Perform a topological sort of the graph using Kahn's algorithm plus a priority queue.
+        # 1. **Initialization**: Create dictionaries to track each node's depth, parent nodes, and original index.
+        # 2. **Queue Initialization**: Track all nodes with zero child nodes in a priority queue, using their original index for ordering.
+        # 3. **Topological Sort**:
+        #   - While the priority queue is not empty:
+        #     - Pop the node with the largest original index and add it to the revesed sorted list.
+        #     - Decrement the depth of its parent nodes. If any parent node's has zero child node, push it into the queue.
+        # 4. **Cycle Check**: Ensure all nodes are processed. If not, raise a `ValueError` indicating a cycle.
+        # 5. **Finishing**: Reverse the sorted list to get the topological order.
+        # Initialization
+        node_depth: dict[Node, int] = dict.fromkeys(self._nodes, 0)
+        node_parents: dict[Node, list[Node]] = {node: [] for node in self._nodes}
+        node_index: dict[Node, int] = {node: i for i, node in enumerate(self._nodes)}
```

Comment:
When a node like `If` has attributes that contains graphs, it may have an implicit dependency on previous nodes that is not seen in the inputs